### DISCUSSION TITLE
Plugin using log and progress leaves progress behind

### DISCUSSION THREAD

**Discussion by coyoteecd** [serverless/serverless#10670] [Q&A]
I am working on a PR for a plugin, adapting it to use the new v3 CLI logging API.
The plugin registers a before:remove:remove action that empties an S3 bucket so that actual removal would not fail.

In my PR, I did what I thought was a nice way of logging:

```js
clear() {
  const taskProgress = this.progress.create({...});
  const promises = [build up promises that do the removal per each configured bucket]
 
  return Promise.all(promises)
    .then(() => {
      this.log.success('It worked');
   })
   .finally(() => {
      taskProgress.remove();
   });
}
```

I tested the change using a serverless config that does **not** have a matching CFN stack and therefore it does not have permissions to describe stack/create changesets etc. So it fails as expected.

Here's what the console log says:
```sh
> serverless remove "--stage" "xxx"

Removing my-stack from stage xxx (us-east-1)
✔ It worked
Environment: win32, node 14.18.2, framework 3.1.1 (local), plugin 6.0.0, SDK 4.3.1
Credentials: Local, "my-profile" profile
Docs:        docs.serverless.com
Support:     forum.serverless.com
Bugs:        github.com/serverless/serverless/issues

Error:
User: arn:redacted is not authorized to perform: cloudformation:DescribeStackResource on resource: arn:redacted/* because no identity-based policy allows the cloudformation:DescribeStackResource action
```

For me, it is reasonable to expect that once progress.remove is called, the CLI line would be gone from the output. I know it's a small detail, so not reporting directly as a bug to see opinions from maintainers.

I have already tried adding the remove right above the log.success line, the behavior is the same.

@medikoo @pgrzesik, thoughts?

**Comment by mnapoli** [comment]
> it is reasonable to expect that once progress.remove is called, **the CLI line would be gone from the output**

Just to clarify, what is the log line that you expect to go away?

If you are talking about `✔ It worked`? -> this line isn't written by the progress, so I'm not sure why you would expect this line to be removed, but maybe I'm misunderstanding.

In any case it's super interesting to see the new API being used and get more feedback on it!

**Comment by medikoo** [comment]
@coyoteecd to further clarify `this.log` writes permanent log messages. While `this.progress.create` returns an interface, which you can use to write dynamic progress info that'll be cleared automatically. And this dynamic info is always presented at the bottom. It's like:

````
1st permanent log message
2nd permanent log message
...
nth permanent log message

Dynamic progress info (will be removed once `remove` is called, or once command processing will end)
```

**Reply by coyoteecd** [reply]
Yes, I understand how it works. Apparently I did not pay attention to the copy-pasted log I posted here, the original output was something like this:

```
Removing stack from stage xxx (us-east-1)

✔ Removed objects from S3 buckets
Removing objects from S3 bucket
Environment: win32, node 14.18.2, framework 3.1.1 (local), plugin 6.0.0, SDK 4.3.1
```

So I had the "Removing..." message (added by progress.create) still there, and was expecting it would be gone.
Now it appears to work properly, weird

**Reply by coyoteecd** [reply]
Also, to give some feedback on usage: what I am trying to achieve is show progress while the plugin is clearing the S3 bucket, and replace the progress with a "success" message once it's done, as a statement in the console that the plugin did its job.

**Reply by medikoo** [reply]
> So I had the "Removing..." message (added by progress.create) still there, and was expecting it would be gone.
Now it appears to work properly, weird

It definitely shouldn't be the case, but you happen to see it again and have reliable reproduction case please post a bug report

**Reply by mnapoli** [reply]
👍  and to complete on this: in the v3 approach, it would make sense to log "It worked" to --verbose output actually. The goal of the default CLI output is to be minimal and only show information that is immediately useful to the user.

If the plugin successfully performs a sub-step of `serverless remove`, then it's an information that doesn't have to be logged to the output because it all went well.

The progress is useful however because the substep may take some time.

So to sum up:

- when the task is happening, and could take a while, it's useful to show the info to the user via a progress (else they may think that something is stuck)
- when the task was done with success, then there is no important info to show to the user anymore and so at this point `progress.remove()` may be enough
  - you could still log to verbose to provide details in case the user runs the command with `--verbose`

Hope that helps!

**Reply by coyoteecd** [reply]
@mnapoli for what is worth, the plugin is https://github.com/k1LoW/serverless-s3-sync. It integrates with deploy/remove, but also has stand-alone commands to sync with an S3 bucket. So the "sync success" messages will actually be the only success messages when running `serverless s3sync` command.

Are you saying that I should just stick to progress and omit the success in this case?

**Reply by mnapoli** [reply]
Ah good point, in case where you have a separate command it's a good thing to show the success message.

When you integrate inside an existing command (`serverless remove`), then the success message of `serverless remove` is enough useful info for the user.

So I admit this creates a more tricky implementation: I've had exactly that problem in the Lift plugin.

You can see what I did here: 

https://github.com/getlift/lift/blob/0d164da931de51b847f4388eb54629b46d9126eb/src/constructs/aws/StaticWebsite.ts#L184-L196

I ended up having 2 methods:

- `uploadWebsiteCommand` -> shows the success message, and is called only for the custom command
- `uploadWebsite` -> doesn't show a success message

Also `uploadWebsiteCommand` calls `uploadWebsite` so all the actual logic is inside `uploadWebsite`. Hope that helps.

**Reply by coyoteecd** [reply]
OK, clear. Yeah I was aware about the possibility to handle it differently, honestly I did not want to go that far because it's not my plugin. I'll consider it, thanks.

One more suggestion: if having one success message per command (be it deploy, remove or custom command), maybe this should be mentioned in the CLI [best practices](https://www.serverless.com/framework/docs/guides/plugins/cli-output). Otherwise people will read "minimal" differently; in my case, I read it as "max one success message from my plugin" when it actually means "max one success message per command".

**Comment by coyoteecd** [comment]
OK, so I got it again and figured out what I was missing initially.
For reference, this is the plugin part that executes at the end, in the `before:remove:remove` hook:

```js
    const taskProgress = this.progress.create({ message: 'Removing objects from S3 buckets' });

    return Promise.all(promises)
      .then(() => {
        console.log('REMOVE CALLED - then');
        this.log.verbose('Removed objects from S3 buckets');
      })
      .finally(() => {
        console.log('REMOVE CALLED - finally');
        taskProgress.remove();
      });
```

When executing the `sls remove`, the progress message gets removed and the verbose log does not show up at all. So, works as intended.
```sh
> serverless remove "--stage" "thor"

Removing stack from stage xxx (us-east-1)
REMOVE CALLED - then
REMOVE CALLED - finally
Environment: win32, node 14.18.2, framework 3.1.1 (local), plugin 6.0.0, SDK 4.3.1
....

Error:
...
```

However when executing with --verbose, I get this:

```sh
> serverless remove "--stage" "thor" "--verbose"

Removing sws-platform-admin from stage thor (us-east-1)

REMOVE CALLED - then
Removed objects from S3 buckets
REMOVE CALLED - finally
Removing objects from S3 bucket       
Environment: win32, node 14.18.2, framework 3.1.1 (local), plugin 6.0.0, SDK 4.3.1
...
Error:
...
```

It looks like verbose flag makes the task permanent for whatever reason. I assume it's not intentional (if it were, at least it would print the **last** message for the progress task, not the one it was created with. I did try to call update and it did not change the output => the message from progress.create was printed.

What is the expected behavior here, should I report a bug?

**Reply by coyoteecd** [reply]
Hmm, I notice that the last char is missing from the progress message, so it might be that it's trying to remove itself, and fails (notice the progress says "buckets" while the output says "bucket")

**Reply by medikoo** [reply]
@coyoteecd do you observe same issue on mac or LInux? Maybe it's some quirk specific to Windows.

Other cause could be, that we have a special internal property for progress, which makes progress messages duplicated in verbose log, still that needs to be explicitly triggered, so it's weird it happens here (but it's another area to explore for potential buggy behavior).

I believe you can report is as bug, as in above case I'd expect any progress messages to be cleared indefinitely

**Reply by coyoteecd** [reply]
@medikoo 
Forgot to mention that I am running it from the PowerShell terminal of VSCode.

I don't have  a Linux around, but for a quick check I tried the Git Bash (on Windows) and it works correctly.
I also tried it from Windows cmd prompt - same issue as in PS.

I'll write a bug for it and mention these details, thanks for your assistance.
